### PR TITLE
Convert net usage to annual nets distributed per capita

### PR DESCRIPTION
**Pull request by norasIDM**
Hi Pete, can you let me know if this is sort of what you had in mind for aim 1? I am also unsure to what degree we want to rely on reading in data from other sources or saving them here.

### CODE REVIEW

File: R/prepare_data.R [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#' Function for reading in and formatting required datasets
+#'
+#' @return 3 datasets of use rate by country, Loess curve of access against nets 
+#' per capita, and estimated median net retention times by country
+#' 
+#' @importFrom utils 'read.csv'
+#' @export
+
+prepare_data <- function() {
```
Comment (pwinskill) [inline R/prepare_data.R]: Can we put all variables that the user might want to change, or that might have updated defaults in the future as arguments for this function.
e.g. data location address strings, year filters etc.

File: R/usage_to_annual_nets_distributed.R [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+#' @export
+
+find_annual_nets_distributed <- function(target_usage, country_iso3="all", 
+                                         extrapolate_npc = "linear", k=20) {
+  
+  ### 1. Convert target usage to nets per capita ###
+  
+  # Read in datasets of country-specific use rate and the Loess curve of 
+  # access vs nets per capita
+  data <- prepare_data()
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: I think it might be best to have this `prepare_data()` call outside of this function then passed through as an argument. Otherwise a user might make lots of individual calls to `find_annual_nets_distributed()` and each time it is downloading and preparingt the whole MAP dataset.

File: R/usage_to_annual_nets_distributed.R [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+#' Bertozzi-Villa, Amelia, et al. Nature communications 12.1 (2021): 1-12.
+#'
+#' @return Dataframe of access, nets per capita and annual nets distributed per capita
+#' for the desired steady-state target usage by country
+#' 
+#' @importFrom stats 'approx' 'reshape' 'loess' 'predict'
+#' @export
+
+find_annual_nets_distributed <- function(target_usage, country_iso3="all", 
+                                         extrapolate_npc = "linear", k=20) {
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: This seems a bit pedantic (sorry!) but good to follow a consistent style guide (e.g. spaces arounf <- and =).
The tidyverse style guide is a good reference here: https://style.tidyverse.org/index.html

File: R/usage_to_annual_nets_distributed.R [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+  
+  max_usage <- country_use_rate[country_use_rate$iso3 %in% country_iso3,c("iso3","use_rate")]
+  names(max_usage)[names(max_usage)=="use_rate"] <- "maximum_usage"
+  
+  if (any(target_usage<=max(country_use_rate[country_use_rate$iso3 %in% country_iso3,"use_rate"]))==FALSE) {
+    print(max_usage)
+    stop("Target usage(s) cannot be achieved with estimated use rate in any of the countries. Refer to maximum usage above.")
+  } else if(any(target_usage>min(country_use_rate[country_use_rate$iso3 %in% country_iso3,"use_rate"]))) {
+    print(max_usage)
+    print("Target usage(s) cannot be achieved with estimated use rate in some of the countries - return NA. Refer to maximum usage above.")
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: Replace these print statements with warnings `warning()` as that will play nicely with formal error checking/handling

File: R/usage_to_annual_nets_distributed.R [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+#' estimate the median net retention times by country. Default = 20 from
+#' Bertozzi-Villa, Amelia, et al. Nature communications 12.1 (2021): 1-12.
+#'
+#' @return Dataframe of access, nets per capita and annual nets distributed per capita
+#' for the desired steady-state target usage by country
+#' 
+#' @importFrom stats 'approx' 'reshape' 'loess' 'predict'
+#' @export
+
+find_annual_nets_distributed <- function(target_usage, country_iso3="all",
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: for consistency can `country_iso3` be renamed `iso3`

File: R/usage_to_annual_nets_distributed.R [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+  # access vs nets per capita
+  data <- prepare_data()
+  country_use_rate <- data$use_rate_by_country
+  loess_for_prediction <- data$loess_for_prediction
+  
+  if(country_iso3[1]=="all") {
+    country_iso3 <- country_use_rate$iso3
+  }
+  
+  if(extrapolate_npc == "linear") {
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: "linear" here makes it seem like there will be other extrapolation options. I'd just have the argument `extrapolate_npc` as boolean, with an explainer in the documetnation that if set to TRUE the extrapolation will be linear

File: R/usage_to_annual_nets_distributed.R [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    # Add datapoint of (1,1) to allow linear extrapolation of nets per capita
+    # for access levels beyond those observed
+    loess_for_prediction <- rbind(loess_for_prediction,
+                                  data.frame(iso3=NA, month=NA, access=1, 
+                                             percapita_nets = 1, loess_predicted_access=1))
+    print("Access-nets per capita curve is extrapolated linearly beyond observed levels")
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R:115]: As below comment (`warning()`)

File: R/usage_to_annual_nets_distributed.R [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+  } else {
+    print("No extrapolation beyond observed access-nets per capita relationship - return NA")
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: Not necessary to have this

File: R/usage_to_annual_nets_distributed.R [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+#' 
+#' Assumptions: 
+#' - Net use rates stays at 2019 levels for all countries
+#' - When converting from access to NPC, use the 2020 loess curve fit of access-NPC 
+#'   at the country-month level.
+#' 
+#' @param target_usage A single value or vector of desired target usages to model.
+#' @param country_iso3 Vector of ISO3 codes of African countries for which to return results. 
+#' Default = "all".
+#' @param extrapolate_npc Option to extrapolate target nets per capita beyond fitted
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: For the extrapolation, I think we want the trend from the end of curve (say the last 4 points) to be extrapolated.

Currently if I run:
```
p1 <- find_annual_nets_distributed(seq(0, 1, 0.01), country_iso3 = "NGA")
plot(p1$annual_percapita_nets_distributed ~ p1$target_use,
     xlab = "Target Use", ylab = "Annual nets distributed")
```
I get:
image

This makes it get more efficient at the end, which I don't think would happen

File: R/usage_to_annual_nets_distributed.R [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+  targets_all$target_access[targets_all$target_access>1] <- NA
+  # Set access to NA where it is over 1 and excluded from conversion
+  
+  ### 2. Convert equilibrium nets per capita to annual nets distributed per capita ###
+  
+  # Read in net half lives (median retention times) for each country
+  half_life_data <- data$half_life_data
+  
+  # Estimate time at which all nets are lost (parameter l) based on net half life in each country
+  half_life_data$l <- half_life_data$half_life/sqrt(1 - k / (k - log(0.5)))
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: This is quite a large function and a number of sections can be split out into their own small functions that perform single, simple tasks. These are then much easier to unit test.

File: R/usage_to_annual_nets_distributed.R [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+                            by = "iso3", all.x=TRUE, all.y=FALSE)
+  nets_distributed$target_usage <- nets_distributed$target_access*nets_distributed$use_rate
+  
+  nets_distributed <- nets_distributed[order(nets_distributed$iso3, nets_distributed$annual_percapita_nets_distributed),]
+  
+  return(nets_distributed[,c("iso3", "use_rate", "annual_percapita_nets_distributed",
+                             "target_percapita_nets","target_access", 
+                             "target_usage")])
+  
+  # TO DO:
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: Good to keep comments like this out of the code - they have a tendancy to sneek through to propoer releases. Feel free to stick stuff like this in the issues + conmversations on those issues.

Extrpolating at the lower end would be helpful yes.
I think we can have percapita nets >1 (its a bit crazy, but not impossible)

File: R/usage_to_annual_nets_distributed.R [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+  prop_retained <- exp(k - k / (1 - (t / l) ^ 2))
+  prop_retained[t >= l] <- 0
+  return(prop_retained)
+}
+
+#' Integrate net loss function between 0 and 3
+#' @param k Fixed rate. Default = 20 from Bertozzi-Villa, Amelia, et al. Nature communications 12.1 (2021): 1-12.
+#' @param l Country-specific time at which no nets are retained (nets=0).
+#' @export
+integrate_net_loss <- function(k,l) {
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: We probably want at least the upper to be an argument here  - we might want to assume other distribution intervals than 3 yearly.

File: R/usage_to_annual_nets_distributed.R [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+                            by = "iso3", all.x=TRUE, all.y=FALSE)
+  
+  # Use net loss function to calculate annual nets distributed per capita for different countries 
+  # and target usages
+  # The annual number of nets distributed to maintain the input equilibrium nets per capita. 
+  # This is for the specific "smoothed" case where we assume that 1/3 of the population at risk
+  # is distributed nets every 3 years (this is equivalent to a smoothed 3 yearly whole 
+  # population distribution).
+  nets_distributed$annual_percapita_nets_distributed <- nets_distributed$target_percapita_nets * 
+    integrate_net_loss_vector(k=k, l = nets_distributed$l) * (1/3)
```
Comment (pwinskill) [inline R/usage_to_annual_nets_distributed.R]: This conversion is assuming equilibrium distributions of equial size on a regular schedule.

Eventually I think we might want to "fit" the annual_percapita_nets_distributed to match the target_per_capita nets over time. This is an (almost) identical operation to the fitting the malariasimulation input distirbutions to match target population usage over time - so we should be able to use the same functionality.

Comment (pwinskill) [reply R/usage_to_annual_nets_distributed.R]: I will add this as an issue

**Review by pwinskill — CHANGES_REQUESTED** [review]
This is going to be so helpful!

Lots of suggestions in specific comments.

Some general suggestions:
1. Lots of the functions remain large. It will be best to break them down as much as possible into small chunks where each function does a single small "job"
2. Doing (1) will also us to include unit tests of these small functions much more easily. We will need to add quite a few tests here to make sure the code is doing what we want - especially as we have quite a few edge cases.

I had some issues with the documentation. Worth making sure you've run `devtools::document()` to update before pushing any commits.

File: R/prepare_data.R (hunk 1 above)
Comment (norasIDM) [reply R/prepare_data.R]: I broke up the functions into the individual conversion steps and added some tests for each. Documentation should be up-to-date

File: R/usage_to_annual_nets_distributed.R (hunk 4 above)
Comment (norasIDM) [reply R/usage_to_annual_nets_distributed.R]: Replaced with warning() where appropriate (some are more information rather than warnings)

File: R/usage_to_annual_nets_distributed.R (hunk 5 above)
Comment (norasIDM) [reply R/usage_to_annual_nets_distributed.R]: Data now read in externally so country selection will also be done at that stage

File: R/usage_to_annual_nets_distributed.R (hunk 6 above)
Comment (norasIDM) [reply R/usage_to_annual_nets_distributed.R]: Added an additional option

File: R/usage_to_annual_nets_distributed.R (hunk 7 above)
Comment (norasIDM) [reply R/usage_to_annual_nets_distributed.R:115]: I did not change this to warning() since it is more of a reminder of what is inputted

File: R/usage_to_annual_nets_distributed.R (hunk 9 above)
Comment (norasIDM) [reply R/usage_to_annual_nets_distributed.R]: I included this unrealistic option in case you wanted to reach an access of 1 (so you can run higher usages) but let me know if you think its better to remove altogether. Option "loess" now added is more realistic
